libs/ledger-live-common/src/deviceSDK/tasks/getLatestFirmware.ts

```diff
@@ -0,0 +1,70 @@
+import type {
+  DeviceId,
+  DeviceInfo,
+  FirmwareUpdateContext,
+} from "@ledgerhq/types-live";
+
+import { quitApp } from "../commands/quitApp";
+
+import { withDevice } from "../../hw/deviceAccess";
+import { from, Observable, of } from "rxjs";
+import { switchMap, catchError } from "rxjs/operators";
+import { SharedTaskEvent, sharedLogicTaskWrapper } from "./core";
+import manager from "../../manager";
+
+export type GetLatestFirmwareTaskArgs = {
+  deviceId: DeviceId;
+  deviceInfo: DeviceInfo;
+};
+
+export type GetLatestFirmwareTaskError =
+  | "FailedToRetrieveFirmwareUpdateInfo"
+  | "FirmwareUpToDate";
+
+export type GetLatestFirmwareTaskErrorEvent = {
+  type: "taskError";
+  error: GetLatestFirmwareTaskError;
+};
+
+export type GetLatestFirmwareTaskEvent =
+  | { type: "data"; firmwareUpdateContext: FirmwareUpdateContext }
+  | GetLatestFirmwareTaskErrorEvent
+  | SharedTaskEvent;
+
+function internalGetLatestFirmwareTask({
+  deviceId,
+  deviceInfo,
+}: GetLatestFirmwareTaskArgs): Observable<GetLatestFirmwareTaskEvent> {
+  return new Observable((subscriber) => {
+    return withDevice(deviceId)((transport) =>
+      quitApp(transport).pipe(
+        switchMap(() => {
+          return from(manager.getLatestFirmwareForDevice(deviceInfo));
+        }),
+        switchMap((firmwareUpdateContext) => {
+          if (firmwareUpdateContext) {
+            return of<GetLatestFirmwareTaskEvent>({
+              type: "data",
+              firmwareUpdateContext,
+            });
+          } else {
+            return of<GetLatestFirmwareTaskEvent>({
+              type: "taskError",
+              error: "FirmwareUpToDate",
```

Comment:
Do we want to return a "taskError" for a firmware that is up to date ? Should it be another type of event ?

Reply:
I think the error makes sense. In theory the consumer would check independently if there's already an update available for the device and only trigger this if yes.

If we make an analogy to HTTP methods, this would be like a 422 (unprocessable entity): "you're asking me to do something that doesn't make sense, to update the firmware of a device that doesn't have any updates"

That was my reasoning.